OpenRA.Game/Network/UPnP.cs

```diff
@@ -10,127 +10,85 @@
 
-		public static void TryNatDiscovery()
-		{
-			try
-			{
-				NatUtility.Logger = Log.Channel("server").Writer;
-				NatUtility.Verbose = Game.Settings.Server.VerboseNatDiscovery;
-				NatUtility.DeviceFound += DeviceFound;
-				Game.Settings.Server.NatDeviceAvailable = false;
-				NatUtility.StartDiscovery();
-				Log.Write("server", "NAT discovery started.");
-			}
-			catch (Exception e)
-			{
-				Log.Write("server", "Can't discover UPnP-enabled device: {0}", e);
-				Game.Settings.Server.NatDeviceAvailable = false;
-				Game.Settings.Server.AllowPortForward = false;
-			}
-		}
+		public static IPAddress ExternalIP;
 
-		public static void StoppingNatDiscovery()
+		public static async Task TryNatDiscovery()
 		{
-			Log.Write("server", "Stopping NAT discovery.");
-			NatUtility.StopDiscovery();
-
-			if (natDevice == null || natDevice.GetType() != typeof(Mono.Nat.Upnp.UpnpNatDevice))
-			{
-				Log.Write("server",
-					"No NAT devices with UPnP enabled found within {0} ms deadline. Disabling automatic port forwarding.".F(Game.Settings.Server.NatDiscoveryTimeout));
-				Game.Settings.Server.NatDeviceAvailable = false;
+			if (Game.Settings.Server.DiscoverNatDevices == false)
 				Game.Settings.Server.AllowPortForward = false;
-			}
-		}
-
-		public static void DeviceFound(object sender, DeviceEventArgs args)
-		{
-			Log.Write("server", "NAT device discovered.");
-
-			Game.Settings.Server.NatDeviceAvailable = true;
-			Game.Settings.Server.AllowPortForward = true;
 
 			try
 			{
-				natDevice = args.Device;
-				Log.Write("server", "Type: {0}", natDevice.GetType());
-				Log.Write("server", "Your external IP is: {0}", natDevice.GetExternalIP());
+				NatDiscoverer.TraceSource.Switch.Level = SourceLevels.Verbose;
+				Log.AddChannel("nat", "nat.log");
+				var logChannel = Log.Channel("nat");
+				NatDiscoverer.TraceSource.Listeners.Add(new TextWriterTraceListener(logChannel.Writer));
 
-				foreach (var mp in natDevice.GetAllMappings())
-					Log.Write("server", "Existing port mapping: protocol={0}, public={1}, private={2}",
-						mp.Protocol, mp.PublicPort, mp.PrivatePort);
+				var natDiscoverer = new NatDiscoverer();
+				var token = new CancellationTokenSource(Game.Settings.Server.NatDiscoveryTimeout);
+				natDevices = await natDiscoverer.DiscoverDevicesAsync(PortMapper.Upnp, token);
+				foreach (var natDevice in natDevices)
+				{
+					try
+					{
+						ExternalIP = natDevice.GetExternalIPAsync().Result;
+					}
+					catch { }
+				}
 			}
 			catch (Exception e)
 			{
-				Log.Write("server", "Can't fetch information from NAT device: {0}", e);
-
-				Game.Settings.Server.NatDeviceAvailable = false;
+				Console.WriteLine("NAT discovery failed: {0}", e.Message);
+				Log.Write("nat", e.StackTrace);
 				Game.Settings.Server.AllowPortForward = false;
 			}
+
+			Game.Settings.Server.AllowPortForward = true;
 		}
 
-		public static void ForwardPort(int lifetime)
+		public static async void ForwardPort()
```

Comment:
Please pass in the ports as arguments instead of reading them from `Game.Settings`.

Passing data around via global state is poor practice, and is why our server and utility classes are full of `// HACK: The engine code assumes that Game.Settings is set.` and `// HACK: The engine code assumes that Game.modData is set.`